### PR TITLE
Tax rates with dates

### PR DESCRIPTION
**Pull request by everwhatever**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         |  1.12                   |
| Bug fix?        | no                                                      |
| New feature?    | yes                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | no  |
| Related tickets |  [jira](https://sylius.atlassian.net/browse/SYL-2277)                    |
| License         | MIT                                                          |

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -51,3 +51,5 @@
 /config/jwt/*.pem
 !/config/jwt/*-test.pem
 ###< lexik/jwt-authentication-bundle ###
+
+.idea
```
Comment (Zales0123) [inline .gitignore]: This should land in your local git exclusion file (`.git/info/exclude`)

File: features/taxation/applying_taxes/applying_correct_taxes_on_visitor_cart_for_specific_date.feature [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        Then my cart total should be "$123.00"
+        And my cart taxes should be "$23.00"
+
+    @ui @api
+    Scenario: Proper taxes for taxed product after specific date
+        Given it is "02-02-2023" now
+        When I add product "PHP T-Shirt" to the cart
+        Then my cart total should be "$115.00"
+        And my cart taxes should be "$15.00"
+
```
Comment (Zales0123) [inline features/taxation/applying_taxes/applying_correct_taxes_on_visitor_cart_for_specific_date.feature]: Extra blank line

File: features/taxation/applying_taxes/applying_correct_taxes_on_visitor_cart_for_specific_date.feature [hunk 3]
```diff
@@ -0,0 +1,32 @@
+@applying_taxes
+Feature: Apply correct taxes on visitor cart for specific date
+    In order to buy goods with correct taxes applied
+    As a Visitor
+    I want to have correct taxes applied to my order
```
Comment (coldic3) [inline features/taxation/applying_taxes/applying_correct_taxes_on_visitor_cart_for_specific_date.feature]:
```suggestion
    I want to have up-to-date taxes applied to my order
```
WDYT?

File: src/Sylius/Bundle/CoreBundle/Migrations/Version20220912091947.php [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+use Doctrine\Migrations\AbstractMigration;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
+final class Version20220912091947 extends AbstractMigration
+{
+    public function getDescription(): string
+    {
+        return '';
```
Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/Migrations/Version20220912091947.php]: We could add some description here

File: src/Sylius/Component/Taxation/Checker/TaxRateDateChecker.php [hunk 5]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Sylius\Component\Taxation\Checker;
+
+use Sylius\Calendar\Provider\DateTimeProviderInterface;
+use Sylius\Component\Taxation\Model\TaxRateInterface;
+
+class TaxRateDateChecker implements TaxRateDateCheckerInterface
```
Comment (Zales0123) [inline src/Sylius/Component/Taxation/Checker/TaxRateDateChecker.php]: Let's make it final

File: src/Sylius/Component/Taxation/Checker/TaxRateDateChecker.php [hunk 6]
```diff
@@ -0,0 +1,32 @@
+<?php
+
```
Comment (Zales0123) [inline src/Sylius/Component/Taxation/Checker/TaxRateDateChecker.php]: Missing license block

File: src/Sylius/Component/Taxation/Checker/TaxRateDateChecker.php [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+use Sylius\Component\Taxation\Model\TaxRateInterface;
+
+class TaxRateDateChecker implements TaxRateDateCheckerInterface
+{
+    public function __construct(
+        protected DateTimeProviderInterface $calendar,
+    ) {
+    }
+
+    public function check(array $taxRates): ?TaxRateInterface
```
Comment (Zales0123) [inline src/Sylius/Component/Taxation/Checker/TaxRateDateChecker.php]: `filter` would be more appropriate, doesn't it? It's what it does 😄

File: src/Sylius/Component/Taxation/Model/TaxRate.php [hunk 8]
```diff
@@ -120,4 +126,48 @@ public function getLabel(): ?string
...
+        return $this->endDate;
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function setEndDate(?DateTimeInterface $endDate): void
+    {
+        if ($this->startDate != null && $endDate < $this->startDate) {
+            throw new Exception("The tax rate should not end before it starts");
```
Comment (Zales0123) [inline src/Sylius/Component/Taxation/Model/TaxRate.php]: I believe this is not speced

File: src/Sylius/Component/Taxation/spec/Checker/TaxRateDateCheckerSpec.php [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+declare(strict_types=1);
+
+namespace spec\Sylius\Component\Taxation\Checker;
+
+use PhpSpec\ObjectBehavior;
+use Sylius\Calendar\Provider\DateTimeProviderInterface;
+use Sylius\Component\Taxation\Model\TaxRateInterface;
+use Sylius\Component\Taxation\Checker\TaxRateDateCheckerInterface;
+
+class TaxRateDateCheckerSpec extends ObjectBehavior
```
Comment (Zales0123) [inline src/Sylius/Component/Taxation/spec/Checker/TaxRateDateCheckerSpec.php]: `final` :dancer:

File: src/Sylius/Component/Taxation/spec/Checker/TaxRateDateCheckerSpec.php [hunk 10]
```diff
@@ -0,0 +1,54 @@
+<?php
+
```
Comment (Zales0123) [inline src/Sylius/Component/Taxation/spec/Checker/TaxRateDateCheckerSpec.php]: Missing license block

File: src/Sylius/Component/Taxation/spec/Checker/TaxRateDateCheckerSpec.php [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+{
+    function let(DateTimeProviderInterface $calendar): void
+    {
+        $this->beConstructedWith($calendar);
+    }
+
+    function it_implements_tax_rate_resolver_interface(): void
+    {
+        $this->shouldImplement(TaxRateDateCheckerInterface::class);
+    }
```
Comment (Zales0123) [inline src/Sylius/Component/Taxation/spec/Checker/TaxRateDateCheckerSpec.php]: Missing blank line below

File: src/Sylius/Component/Taxation/Resolver/TaxRateResolver.php [hunk 12]
```diff
@@ -13,14 +13,18 @@
...
 use Sylius\Component\Taxation\Model\TaxableInterface;
 use Sylius\Component\Taxation\Model\TaxRateInterface;
 
 class TaxRateResolver implements TaxRateResolverInterface
 {
-    public function __construct(protected RepositoryInterface $taxRateRepository)
-    {
+    public function __construct(
+        protected RepositoryInterface $taxRateRepository,
+        protected TaxRateDateCheckerInterface $taxRateDateChecker
```
Comment (Zales0123) [inline src/Sylius/Component/Taxation/Resolver/TaxRateResolver.php]: To provide a BC layer, we need to make it nullable and check in the `resolve` method. Also throw a deprecation when it is not passed

File: src/Sylius/Behat/Context/Ui/Admin/ManagingTaxRateContext.php [hunk 13]
```diff
@@ -13,6 +13,7 @@
 
 namespace Sylius\Behat\Context\Ui\Admin;
 
+use Behat\Behat\Tester\Exception\PendingException;
```
Comment (coldic3) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingTaxRateContext.php]: Is it used?

File: features/taxation/managing_tax_rates/adding_tax_rate.feature [hunk 14]
```diff
@@ -35,6 +35,20 @@ Feature: Adding a new tax rate
         Then I should be notified that it has been successfully created
         And the tax rate "United States Sales Tax" should appear in the registry
 
+    @ui
+    Scenario: Adding a new tax rate with start and end date
```
Comment (coldic3) [inline features/taxation/managing_tax_rates/adding_tax_rate.feature:39]: 👍️ I would also test the start date without the end date and vice versa.

File: src/Sylius/Bundle/CoreBundle/Migrations/Version20220912091947.php [hunk 15]
```diff
@@ -0,0 +1,31 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\Migrations;
+
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
```
Comment (coldic3) [inline src/Sylius/Bundle/CoreBundle/Migrations/Version20220912091947.php]: We should use `Sylius\Bundle\CoreBundle\Doctrine\Migrations\AbstractMigration` instead to restrict the migration to MySQL only.

File: src/Sylius/Component/Taxation/Checker/TaxRateDateChecker.php [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+        /** @var TaxRateInterface $taxRate */
+        foreach ($taxRates as $key => $taxRate) {
+            if (!$taxRate->isInDate($now)) {
+                unset($taxRates[$key]);
+            }
+        }
+
+        $taxRates = array_values($taxRates);
+
+        return (empty($taxRates)) ? null : $taxRates[0];
```
Comment (coldic3) [inline src/Sylius/Component/Taxation/Checker/TaxRateDateChecker.php]:
```suggestion
        return $taxRates[0] ?? null;
```

File: src/Sylius/Component/Taxation/Model/TaxRate.php [hunk 17]
```diff
@@ -120,4 +126,48 @@ public function getLabel(): ?string
...
+    {
+        return $this->endDate;
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function setEndDate(?DateTimeInterface $endDate): void
+    {
+        if ($this->startDate != null && $endDate < $this->startDate) {
```
Comment (coldic3) [inline src/Sylius/Component/Taxation/Model/TaxRate.php]: I may be wrong but I think we should use Validation Constraint to validate it.

File: src/Sylius/Component/Taxation/Model/TaxRate.php [hunk 18]
```diff
@@ -120,4 +126,48 @@ public function getLabel(): ?string
...
+
+        if (null === $this->startDate) {
+            return $date <= $this->endDate;
+        }
+
+        if (null === $this->endDate) {
+            return $date >= $this->startDate;
+        }
+
+        return $this->startDate <= $date && $this->endDate >= $date;
```
Comment (coldic3) [inline src/Sylius/Component/Taxation/Model/TaxRate.php]:
```suggestion
        return $this->startDate <= $date && $this->endDate >= $date;
```

I belive it will work as well, but maybe I'm wrong.

File: src/Sylius/Component/Taxation/Model/TaxRateInterface.php [hunk 19]
```diff
@@ -42,4 +42,14 @@ public function getCalculator(): ?string;
...
+
+    public function getStartDate(): ?\DateTimeInterface;
+
+    public function setStartDate(?\DateTimeInterface $startDate): void;
+
+    public function getEndDate(): ?\DateTimeInterface;
+
+    public function setEndDate(?\DateTimeInterface $endDate): void;
+
+    public function isInDate(\DateTimeInterface $date): bool;
```
Comment (coldic3) [inline src/Sylius/Component/Taxation/Model/TaxRateInterface.php]: I have mixed feelings about this method in general. IMO it's not a part of the model as we have to pass the `$date` argument to it.

**Comment by coldic3** [comment]
Could you also remove `composer.phar` and `env.local`?

File: src/Sylius/Behat/Context/Setup/TaxationContext.php [hunk 20]
```diff
@@ -50,25 +50,33 @@ public function storeHasTaxRateWithinZone(
...
+        string $taxCategoryName,
+        ZoneInterface $zone,
+        string $endDate
+    ) {
+        $this->configureTaxRate($taxCategoryName, null, $taxRateName, $zone, $taxRateAmount, false, null, new \DateTime($endDate));
+    }
 
-        $this->sharedStorage->set('tax_rate', $taxRate);
+    /**
+     * @Given the store has :taxRateName tax rate of :taxRateAmount% for :taxCategoryName within the :zone zone starting at :endDate
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Context/Setup/TaxationContext.php]:
```suggestion
     * @Given the store has :taxRateName tax rate of :taxRateAmount% for :taxCategoryName within the :zone zone starting at :startDate
```

File: src/Sylius/Behat/Context/Ui/Admin/ManagingTaxRateContext.php (hunk 13 above)
Comment (everwhatever) [reply src/Sylius/Behat/Context/Ui/Admin/ManagingTaxRateContext.php]: no, removed

File: src/Sylius/Component/Taxation/Model/TaxRate.php (hunk 18 above)
Comment (everwhatever) [reply src/Sylius/Component/Taxation/Model/TaxRate.php]: It has to be the way it is

File: .gitignore (hunk 1 above)
Comment (vvasiloi) [reply .gitignore]: Better in a global ignore file. 
https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files#configuring-ignored-files-for-all-repositories-on-your-computer

**Comment by vvasiloi** [comment]
Can this be reduced to a custom query?
Are repository methods part of the BC promise?

File: src/Sylius/Bundle/CoreBundle/Migrations/Version20220912091947.php [hunk 21]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\Migrations;
+
+use Doctrine\DBAL\Schema\Schema;
+use Sylius\Bundle\CoreBundle\Doctrine\Migrations\AbstractMigration;
+/**
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/CoreBundle/Migrations/Version20220912091947.php]:
```suggestion
use Sylius\Bundle\CoreBundle\Doctrine\Migrations\AbstractMigration;

/**
```

File: src/Sylius/Behat/Context/Setup/TaxationContext.php [hunk 22]
```diff
@@ -196,4 +204,42 @@ private function getCodeFromNameAndZoneCode($taxRateName, $zoneCode)
...
+        $taxRate->setCategory($taxCategory);
+        $taxRate->setCalculator('default');
+        $taxRate->setIncludedInPrice($includedInPrice);
+
+        if (null !== $startDate) {
+            $taxRate->setStartDate($startDate);
+        }
+        if (null !== $endDate) {
+            $taxRate->setEndDate($endDate);
+        }
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Context/Setup/TaxationContext.php]: There's no point to these conditions, both methods accept null

File: src/Sylius/Component/Taxation/Checker/TaxRateDateChecker.php [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+
+        if (null === $startDate) {
+            return $date <= $endDate;
+        }
+
+        if (null === $endDate) {
+            return $date >= $startDate;
+        }
+
+        return $startDate <= $date && $endDate >= $date;
```
Comment (NoResponseMate) [inline src/Sylius/Component/Taxation/Checker/TaxRateDateChecker.php]:
```suggestion
        if (null === $endDate) {
            return $startDate <= $date;
        }

        return $startDate <= $date && $endDate >= $date;
```
I think it should be enough to take care of all the cases

File: src/Sylius/Component/Taxation/Checker/TaxRateDateCheckerInterface.php [hunk 24]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+declare(strict_types=1);
```
Comment (NoResponseMate) [inline src/Sylius/Component/Taxation/Checker/TaxRateDateCheckerInterface.php]: Missing license

**Comment by vvasiloi** [comment]
> Can this be reduced to a custom query? Are repository methods part of the BC promise?

There isn't a custom repository for tax rates, so there won't be a BC break.
The repository could look like this:
```php
class TaxRateRepository extends EntityRepository implements TaxRateRepositoryInterface
{
    public function findOneByCategoryAndDate(TaxCategoryInterface $taxCategory, ?\DateTimeInterface $date = null): ?TaxRateInterface {
        if (null === $date) {
            $date = new \DateTime();
        }

        $expr = $this->_em->getExpressionBuilder();

        return $this->createQueryBuilder('o')
            ->where(
                $expr->andX(
                    $expr->eq('o.category', ':category'),
                    $expr->orX(
                        $expr->isNull('o.startDate'),
                        $expr->lte('o.startDate', ':date')
                    ),
                    $expr->orX(
                        $expr->isNull('o.endDate'),
                        $expr->gte('o.endDate', ':date')
                    ),
                )
            )
            ->setParameter('category', $taxCategory)
            ->setParameter('date', $date)
            ->getQuery()
            ->getOneOrNullResult();
    }
}
```

However, a way to add the zone in the core bundle is needed, which complicates things a bit. 🤔 
We could add a child repo in the core bundle with another method that also takes in the zone and extracts the common query builder/expression, but the fact that the resolver accepts a criteria array that's passed directly to the `findBy`/`findOneBy` makes things weird.

If we keep the date filter logic in PHP, then I suggest transforming the `TaxRateDateChecker` into an "eligibility checker" (which is common concept in Sylius) that will take a single tax rate and return true if it fulfils the conditions and false otherwise.
Basically, the checker will contain the logic that's now in `TaxRateDateChecker::isInDate`, which now is a public method that's not part of the abstraction.
Then the resolver will iterate over the tax rates returned from the repository and return the first one that's eligible.

File: src/Sylius/Component/Taxation/Checker/TaxRateDateChecker.php [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+                unset($taxRates[$key]);
+            }
+        }
+
+        $taxRates = array_values($taxRates);
+
+        return $taxRates[0] ?? null;
+    }
+
+    public function isInDate(DateTimeInterface $date, ?DateTimeInterface $startDate, ?DateTimeInterface $endDate): bool
```
Comment (NoResponseMate) [inline src/Sylius/Component/Taxation/Checker/TaxRateDateChecker.php]:
```suggestion
    private function isInDate(TaxRateInterface $taxRate, DateTimeInterface $date): bool
```

File: src/Sylius/Component/Taxation/Checker/TaxRateDateChecker.php [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+    public function filter(array $taxRates): ?TaxRateInterface
+    {
+        $now = $this->calendar->now();
+
+        /** @var TaxRateInterface $taxRate */
+        foreach ($taxRates as $key => $taxRate) {
+            if (!$this->isInDate($now, $taxRate->getStartDate(), $taxRate->getEndDate())) {
+                unset($taxRates[$key]);
+            }
+        }
```
Comment (NoResponseMate) [inline src/Sylius/Component/Taxation/Checker/TaxRateDateChecker.php]: Can be replaced by `array_filter`

**Review by NoResponseMate — COMMENTED** [review]
Missing behat tests for tax rate form validation

File: src/Sylius/Component/Taxation/Checker/TaxRateDateChecker.php [hunk 27]
```diff
@@ -0,0 +1,51 @@
...
+    {
+        $taxRates = array_filter($taxRates, function ($taxRate){
+            if ($this->isInDate($this->calendar->now(), $taxRate)) {
+                return $taxRate;
+            }
+        });
+
+        $taxRates = array_values($taxRates);
+
+        return $taxRates[0] ?? null;
```
Comment (vvasiloi) [inline src/Sylius/Component/Taxation/Checker/TaxRateDateChecker.php]: Just replacing the foreach with an array_filter is not an improvement, especially when doing it wrong (the callable should return true/false, not objects).
An improvement would be to bring back the foreach, but with an early return.
That's because there's no point to iterate over all tax rates, you only need the first one that's eligible.
```suggestion
        foreach ($taxRates as $taxRate) {
            if ($this->isInDate($this->calendar->now(), $taxRate)) {
                return $taxRate;
            }
        }

        return null;
```

File: src/Sylius/Component/Taxation/Model/TaxRate.php (hunk 18 above)
Comment (everwhatever) [reply src/Sylius/Component/Taxation/Model/TaxRate.php]: it has been changed to shorter version

File: features/taxation/applying_taxes/applying_correct_taxes_on_visitor_cart_for_specific_date.feature [hunk 28]
```diff
@@ -0,0 +1,31 @@
+@applying_taxes
+Feature: Apply correct taxes on visitor cart for specific date
```
Comment (GSadee) [inline features/taxation/applying_taxes/applying_correct_taxes_on_visitor_cart_for_specific_date.feature]:
```suggestion
Feature: Applying correct taxes on visitor cart for a specific date
```

File: features/taxation/applying_taxes/applying_correct_taxes_on_visitor_cart_for_specific_date.feature [hunk 29]
```diff
@@ -0,0 +1,31 @@
...
+        And the store ships everywhere for Free
+        And default tax zone is "US"
+        And the store has "RoW VAT" tax rate of 10% for "Clothes" for the rest of the world
+        And the store has "US VAT (2022)" tax rate of 23% for "Clothes" within the "US" zone ending at "31-12-2022"
+        And the store has "US VAT (2023-)" tax rate of 15% for "Clothes" within the "US" zone starting at "01-01-2023"
+        And the store has a product "PHP T-Shirt" priced at "$100.00"
+        And it belongs to "Clothes" tax category
+
+    @ui @api
+    Scenario: Proper taxes for taxed product
```
Comment (GSadee) [inline features/taxation/applying_taxes/applying_correct_taxes_on_visitor_cart_for_specific_date.feature]:
```suggestion
    Scenario: Applying proper taxes for product
```

File: features/taxation/applying_taxes/applying_correct_taxes_on_visitor_cart_for_specific_date.feature [hunk 30]
```diff
@@ -0,0 +1,31 @@
...
+        Given it is "01-11-2022" now
+        When I add product "PHP T-Shirt" to the cart
+        Then my cart total should be "$123.00"
+        And my cart taxes should be "$23.00"
+
+    @ui @api
+    Scenario: Proper taxes for taxed product after specific date
```
Comment (GSadee) [inline features/taxation/applying_taxes/applying_correct_taxes_on_visitor_cart_for_specific_date.feature]:
```suggestion
    Scenario: Applying proper taxes for product
```

File: src/Sylius/Bundle/CoreBundle/Migrations/Version20220912091947.php [hunk 31]
```diff
@@ -0,0 +1,31 @@
+<?php
+
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Migrations/Version20220912091947.php:2]: Add license doc block

File: src/Sylius/Component/Taxation/Checker/TaxRateDateEligibilityChecker.php [hunk 32]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Component\Taxation\Checker;
+
+use DateTimeInterface;
```
Comment (GSadee) [inline src/Sylius/Component/Taxation/Checker/TaxRateDateEligibilityChecker.php]:
```suggestion
```

File: src/Sylius/Component/Taxation/Model/TaxRate.php [hunk 33]
```diff
@@ -13,6 +13,8 @@
 
 namespace Sylius\Component\Taxation\Model;
 
+use DateTimeInterface;
+use Exception;
```
Comment (GSadee) [inline src/Sylius/Component/Taxation/Model/TaxRate.php]:
```suggestion
```

File: src/Sylius/Component/Taxation/Model/TaxRate.php [hunk 34]
```diff
@@ -40,6 +42,10 @@ class TaxRate implements TaxRateInterface
     /** @var string|null */
     protected $calculator;
 
+    protected ?DateTimeInterface $startDate = null;
```
Comment (GSadee) [inline src/Sylius/Component/Taxation/Model/TaxRate.php]:
```suggestion
    protected ?\DateTimeInterface $startDate = null;
```
and in other places

File: src/Sylius/Component/Taxation/Model/TaxRateInterface.php [hunk 35]
```diff
@@ -13,6 +13,7 @@
 
 namespace Sylius\Component\Taxation\Model;
 
+use DateTimeInterface;
```
Comment (GSadee) [inline src/Sylius/Component/Taxation/Model/TaxRateInterface.php]:
```suggestion
```

File: src/Sylius/Component/Taxation/Resolver/TaxRateResolver.php [hunk 36]
```diff
@@ -13,14 +13,18 @@
...
 use Sylius\Component\Taxation\Model\TaxableInterface;
 use Sylius\Component\Taxation\Model\TaxRateInterface;
 
 class TaxRateResolver implements TaxRateResolverInterface
 {
-    public function __construct(protected RepositoryInterface $taxRateRepository)
-    {
+    public function __construct(
+        protected RepositoryInterface $taxRateRepository,
+        protected ?TaxRateDateEligibilityCheckerInterface $taxRateDateChecker = null
```
Comment (GSadee) [inline src/Sylius/Component/Taxation/Resolver/TaxRateResolver.php:25]: Add a note to UPGRADE file ([example](https://github.com/Sylius/Sylius/blob/1.12/UPGRADE-1.10.md#upgrade-from-v1108-to-v11010)) about this change and trigger deprecation if there is no second argument passed ([example](https://github.com/Sylius/Sylius/blob/1.12/src/Sylius/Component/Core/Cart/Context/ShopBasedCartContext.php#L48:L50)).

File: src/Sylius/Component/Taxation/Resolver/TaxRateResolver.php [hunk 37]
```diff
@@ -31,6 +35,23 @@ public function resolve(TaxableInterface $taxable, array $criteria = []): ?TaxRa
...
+        if ($this->taxRateDateChecker) {
+            $taxRates = $this->taxRateRepository->findBy($criteria);
+
+            return $this->checkFetchedTaxRates($taxRates);
+        }
+
         return $this->taxRateRepository->findOneBy($criteria);
     }
+
+    private function checkFetchedTaxRates(array $taxRates): ?TaxRateInterface
```
Comment (GSadee) [inline src/Sylius/Component/Taxation/Resolver/TaxRateResolver.php]:
```suggestion
    private function provideEligibleTaxRate(array $taxRates): ?TaxRateInterface
```

File: src/Sylius/Component/Taxation/spec/Checker/TaxRateDateEligibilityCheckerSpec.php [hunk 38]
```diff
@@ -0,0 +1,73 @@
...
+    {
+        $this->shouldImplement(TaxRateDateEligibilityCheckerInterface::class);
+    }
+
+    function it_can_be_in_date_when_both_dates_are_defined(
+        TaxRateInterface $taxRate1,
+        TaxRateInterface $taxRate2,
+        DateTimeProviderInterface $calendar
+    ): void
+    {
```
Comment (GSadee) [inline src/Sylius/Component/Taxation/spec/Checker/TaxRateDateEligibilityCheckerSpec.php]:
```suggestion
    ): void {
```

File: src/Sylius/Component/Taxation/spec/Resolver/TaxRateResolverSpec.php [hunk 39]
```diff
@@ -33,26 +34,26 @@ function it_implements_tax_rate_resolver_interface(): void
...
-        TaxableInterface $taxable,
-        TaxCategoryInterface $taxCategory,
-        TaxRateInterface $taxRate,
+        RepositoryInterface                    $taxRateRepository,
+        TaxRateDateEligibilityCheckerInterface $taxRateDateChecker,
+        TaxableInterface                       $taxable,
+        TaxCategoryInterface                   $taxCategory,
+        TaxRateInterface                       $firstTaxRate,
+        TaxRateInterface                       $secondTaxRate,
+        TaxRateInterface                       $thirdTaxRate
```
Comment (GSadee) [inline src/Sylius/Component/Taxation/spec/Resolver/TaxRateResolverSpec.php]: Do not align arguments 🙏🏻

File: src/Sylius/Component/Taxation/spec/Checker/TaxRateDateEligibilityCheckerSpec.php [hunk 40]
```diff
@@ -0,0 +1,73 @@
...
+        $taxRate1->getEndDate()->willReturn(null);
+
+        $startDate2 = new DateTime('21-09-2029');
+        $taxRate2->getStartDate()->willReturn($startDate2);
+        $taxRate2->getEndDate()->willReturn(null);
+
+        $this->isEligible($taxRate1)->shouldReturn(true);
+        $this->isEligible($taxRate2)->shouldReturn(false);
+    }
+}
```
Comment (GSadee) [inline src/Sylius/Component/Taxation/spec/Checker/TaxRateDateEligibilityCheckerSpec.php:91]: Add also a spec for a case when the start date is null

File: src/Sylius/Component/Taxation/Model/TaxRate.php (hunk 34 above)
Comment (everwhatever) [reply src/Sylius/Component/Taxation/Model/TaxRate.php]: Why should I remove "use" and instead use backslash with DateTimeInterface?

File: src/Sylius/Component/Taxation/Checker/TaxRateDateEligibilityChecker.php [hunk 41]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        $date = $this->calendar->now();
+        $startDate = $taxRate->getStartDate();
+        $endDate = $taxRate->getEndDate();
+
+        if (null === $endDate) {
+            return $startDate <= $date;
+        }
+
+        return $startDate <= $date && $endDate >= $date;
```
Comment (vvasiloi) [inline src/Sylius/Component/Taxation/Checker/TaxRateDateEligibilityChecker.php]: This can be a one-liner (or two if it exceeds the limit).
It will also avoid comparing null to DateTime when `$startDate` is null.
```suggestion
        return  (null === $startDate || $startDate <= $date) && (null === $endDate || $endDate >= $date);
```

File: src/Sylius/Component/Taxation/Model/TaxRate.php (hunk 34 above)
Comment (GSadee) [reply src/Sylius/Component/Taxation/Model/TaxRate.php]: I don't have a perfect answer except that we are using DateTimeInterface in this way in Sylius and it would be consistent with other services. However, it is not a blocker for this PR of course 😃

**Comment by vvasiloi** [comment]
🤔 maybe we can avoid the deprecation by decorating the original resolver instead of modifying it? WDYT?

Edit:
- Add an EligibleTaxRateResolver next to the current TaxRateResolver in the taxation component
- Add the decoration to the taxation bundle

**Comment by Zales0123** [comment]
I believe that the deprecation is not a bad option here - of course, it can cause _discomfort_ for the service user, but on the other hand, we transparently inform about important changes in tax rates resolving logic :rocket: Moreover, as the interface allows to return only one tax rate, we could not use the result of the base resolver to filter the ineligible tax rates, as we need an array of them 🖖

**Comment by vvasiloi** [comment]
@Zales0123 my idea was to completely ignore the output from the current resolver, as it remains only for compatibility for those using the service somehow else. Now that I said it out loud it doesn't sound that good anymore.

**Review by GSadee — APPROVED** [review]

**Review by NoResponseMate — APPROVED** [review]

File: src/Sylius/Bundle/TaxationBundle/Resources/config/services.xml [hunk 42]
```diff
@@ -39,7 +39,13 @@
 
         <service id="sylius.tax_rate_resolver" class="Sylius\Component\Taxation\Resolver\TaxRateResolver">
             <argument type="service" id="sylius.repository.tax_rate" />
+            <argument type="service" id="sylius.tax_rate_date_eligibility_checker" on-invalid="null" />
```
Comment (coldic3) [inline src/Sylius/Bundle/TaxationBundle/Resources/config/services.xml:42]:
```suggestion
            <argument type="service" id="sylius.tax_rate_date_eligibility_checker" />
```
I guess it's not needed.

**Review by coldic3 — APPROVED** [review]

**Review by Zales0123 — APPROVED** [review]

**Comment by Zales0123** [comment]
Thanks, @everwhatever! :tada:
